Sources/Data/Attributes.swift

```diff
@@ -0,0 +1,299 @@
+        func encodeToContainer(dictionary: [String: Any], container: inout KeyedEncodingContainer<Attributes.DynamicCodingKeys>) throws {
+            try dictionary.forEach { (codingKey, value) in
+                let key = DynamicCodingKeys(stringValue: codingKey)
+                switch value {
+                case let value as Int:
+                    try container.encode(value, forKey: key)
+                case let value as Bool:
+                    try container.encode(value, forKey: key)
+                case let value as String:
+                    try container.encode(value, forKey: key)
+                case let value as Double:
+                    try container.encode(value, forKey: key)
+                case let array as [Any]:
+                    var arrayContainer = container.nestedUnkeyedContainer(forKey: key)
+                    try array.forEach({ (item) in
+                        // sadly must duplicate the primitive value type coercion from above, since KeyedEncodingContainer and UnkeyedEncodingContainer have different interfaces.
+                        switch item {
+                        case let value as Int:
+                            try arrayContainer.encode(value)
+                        case let value as Bool:
+                            try arrayContainer.encode(value)
+                        case let value as String:
+                            try arrayContainer.encode(value)
+                        case let value as Double:
+                            try arrayContainer.encode(value)
+                        default:
+                            let context = EncodingError.Context(codingPath: arrayContainer.codingPath, debugDescription: "Expected one of Int, String, Double, Boolean. Rover attributes arrays may only contain those primitive types.")
+                            throw EncodingError.invalidValue(item, context)
+                        }
+                    })
+                case let value as Dictionary<String, Any>:
+                    var nestedContainer = container.nestedContainer(keyedBy: DynamicCodingKeys.self, forKey: key)
+                    try encodeToContainer(dictionary: value, container: &nestedContainer)
+                default:
+                    let context = EncodingError.Context(codingPath: container.codingPath, debugDescription: "Unexpected attribute value")
+                    throw EncodingError.invalidValue(value, context)
+                }
+            }
+        }
+        var container = encoder.container(keyedBy: DynamicCodingKeys.self)
+
+        try encodeToContainer(dictionary: self.rawValue, container: &container)
+    }
+    
+    override init() {
+        rawValue = [:]
+        super.init()
+    }
+}
+
+
+typealias AttributeValue = [String: Any]
+
+extension Dictionary where Key == String, Value: Any {
+    fileprivate var asAttributesForNsDictionary: [String: Any] {
```

Comment:
If you move this into the `Attributes` initializer, you will only need to reference this in one place. The `var attributes: NSDictionary` can be moved into the `func encode(with aCoder: NSCoder)` at which point you know you the `rawValue` property is guaranteed to already contain a valid dictionary.